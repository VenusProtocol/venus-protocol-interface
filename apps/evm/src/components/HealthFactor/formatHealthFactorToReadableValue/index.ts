import BigNumber from 'bignumber.js';
import shortenValueWithSuffix from 'utilities/shortenValueWithSuffix';

export const HEALTH_FACTOR_MAX_VALUE = 100;
export const HEALTH_FACTOR_MAX_DECIMALS = 2;

export const formatHealthFactorToReadableValue = ({ value }: { value: number }) => {
  let readableValue: string;

  if (value > HEALTH_FACTOR_MAX_VALUE) {
    readableValue = `${'> '}${shortenValueWithSuffix({
      value: new BigNumber(HEALTH_FACTOR_MAX_VALUE),
      maxDecimalPlaces: HEALTH_FACTOR_MAX_DECIMALS,
    })}`;
  } else {
    readableValue = shortenValueWithSuffix({
      value: new BigNumber(value),
      maxDecimalPlaces: HEALTH_FACTOR_MAX_DECIMALS,
    });
  }

  return readableValue;
};
