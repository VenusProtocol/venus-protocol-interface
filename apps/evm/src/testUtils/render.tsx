import { QueryClient, QueryClientProvider } from '@tanstack/react-query';
import { render as renderComponentTl } from '@testing-library/react';
import { renderHook as renderHookTl } from '@testing-library/react-hooks';
import type { ReactElement } from 'react';
import { MemoryRouter, Route, Routes } from 'react-router-dom';
import type Vi from 'vitest';

import fakeSigner from '__mocks__/models/signer';

import { Web3Wrapper, useAccountAddress, useChainId, useSigner } from 'libs/wallet';
import { MuiThemeProvider } from 'theme/MuiThemeProvider';
import type { ChainId } from 'types';

const createQueryClient = () =>
  new QueryClient({
    defaultOptions: {
      queries: {
        retry: false,
        gcTime: 0,
      },
    },
  });

interface Options {
  accountAddress?: string;
  chainId?: ChainId;
  routerInitialEntries?: string[];
  routePath?: string;
}

interface WrapperProps {
  queryClient: QueryClient;
  children?: React.ReactNode;
  options?: Partial<Options>;
}

const Wrapper: React.FC<WrapperProps> = ({ children, queryClient, options }) => {
  if (options?.accountAddress) {
    const accountAddress = options?.accountAddress;

    (useAccountAddress as Vi.Mock).mockImplementation(() => ({
      accountAddress,
    }));

    (useSigner as Vi.Mock).mockImplementation(() => ({
      signer: {
        ...fakeSigner,
        getAddress: async () => accountAddress,
      },
    }));
  }

  if (options?.chainId) {
    (useChainId as Vi.Mock).mockImplementation(() => ({
      chainId: options?.chainId,
    }));
  }

  return (
    <MuiThemeProvider>
      <QueryClientProvider client={queryClient}>
        <Web3Wrapper>
          <MemoryRouter initialEntries={options?.routerInitialEntries || ['/']}>
            <Routes>
              <Route path={options?.routePath || '/'} element={children} />
            </Routes>
          </MemoryRouter>
        </Web3Wrapper>
      </QueryClientProvider>
    </MuiThemeProvider>
  );
};

export const renderComponent = (children: ReactElement, options?: Partial<Options>) => {
  const queryClient = createQueryClient();

  const renderRes = renderComponentTl(children, {
    wrapper: props => <Wrapper queryClient={queryClient} options={options} {...props} />,
  });

  return {
    ...renderRes,
    queryClient,
  };
};

export const renderHook = <TProps, TResult>(
  hook: (props: TProps) => TResult,
  options?: Partial<Options>,
) => {
  const queryClient = createQueryClient();

  const renderRes = renderHookTl(hook, {
    wrapper: props => <Wrapper queryClient={queryClient} options={options} {...props} />,
  });

  return {
    ...renderRes,
    queryClient,
  };
};
