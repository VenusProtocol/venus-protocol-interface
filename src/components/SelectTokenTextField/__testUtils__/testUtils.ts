// eslint-disable-next-line import/no-extraneous-dependencies
import { fireEvent, getByTestId } from '@testing-library/react';
import { Token } from 'types';

import { getTokenListItemTestId, getTokenSelectButtonTestId } from '../testIdGetters';

export const selectToken = ({
  token,
  selectTokenTextFieldTestId,
  container,
}: {
  token: Token;
  selectTokenTextFieldTestId: string;
  container: HTMLElement;
}) => {
  // Open token list
  fireEvent.click(
    getByTestId(
      container,
      getTokenSelectButtonTestId({ parentTestId: selectTokenTextFieldTestId }),
    ),
  );

  // Select token
  fireEvent.click(
    getByTestId(
      container,
      getTokenListItemTestId({
        parentTestId: selectTokenTextFieldTestId,
        tokenAddress: token.address,
      }),
    ),
  );
};
