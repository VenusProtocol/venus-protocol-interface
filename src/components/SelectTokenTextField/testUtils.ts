import { fireEvent, getByTestId } from '@testing-library/react';
import { Token } from 'types';

import TEST_ID_FRAGMENTS from './testIdFragments';

export const getTokenInput = ({
  selectTokenTextFieldTestId,
  container,
}: {
  selectTokenTextFieldTestId: string;
  container: HTMLElement;
}) => getByTestId(container, `${selectTokenTextFieldTestId}${TEST_ID_FRAGMENTS.tokenTextField}`);

export const getTokenSelectButton = ({
  selectTokenTextFieldTestId,
  container,
}: {
  selectTokenTextFieldTestId: string;
  container: HTMLElement;
}) => getByTestId(container, `${selectTokenTextFieldTestId}${TEST_ID_FRAGMENTS.tokenSelectButton}`);

export const selectToken = ({
  token,
  selectTokenTextFieldTestId,
  container,
}: {
  token: Token;
  selectTokenTextFieldTestId: string;
  container: HTMLElement;
}) => {
  // Open token list
  fireEvent.click(getTokenSelectButton({ container, selectTokenTextFieldTestId }));

  // Select token
  fireEvent.click(
    getByTestId(
      container,
      `${selectTokenTextFieldTestId}${TEST_ID_FRAGMENTS.tokenListItem}-${token.address}`,
    ),
  );
};
