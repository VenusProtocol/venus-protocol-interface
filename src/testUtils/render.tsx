import { render as renderComponentTl } from '@testing-library/react';
import { renderHook as renderHookTl } from '@testing-library/react-hooks';
import { ReactElement } from 'react';
import { QueryClient, QueryClientProvider } from 'react-query';
import { MemoryRouter, Route, Routes } from 'react-router-dom';
import Vi from 'vitest';

import fakeSigner from '__mocks__/models/signer';

import { Web3Wrapper, useAccountAddress, useChainId, useSigner } from 'packages/wallet';
import { MuiThemeProvider } from 'theme/MuiThemeProvider';
import { ChainId } from 'types';

const createQueryClient = () =>
  new QueryClient({
    defaultOptions: {
      queries: {
        retry: false,
        cacheTime: 0,
      },
    },
  });

interface Options {
  accountAddress?: string;
  chainId?: ChainId;
  routerInitialEntries?: string[];
  routePath?: string;
}

interface WrapperProps {
  queryClient: QueryClient;
  children?: React.ReactNode;
  options?: Options;
}

const Wrapper: React.FC<WrapperProps> = ({ children, queryClient, options }) => {
  if (options?.accountAddress) {
    const accountAddress = options?.accountAddress;

    (useAccountAddress as Vi.Mock).mockImplementation(() => ({
      accountAddress,
    }));

    (useSigner as Vi.Mock).mockImplementation(() => ({
      signer: {
        ...fakeSigner,
        getAddress: async () => accountAddress,
      },
    }));
  }

  if (options?.chainId) {
    (useChainId as Vi.Mock).mockImplementation(() => ({
      chainId: options?.chainId,
    }));
  }

  return (
    <MuiThemeProvider>
      <QueryClientProvider client={queryClient}>
        <Web3Wrapper>
          <MemoryRouter initialEntries={options?.routerInitialEntries || ['/']}>
            <Routes>
              <Route path={options?.routePath || '/'} element={children} />
            </Routes>
          </MemoryRouter>
        </Web3Wrapper>
      </QueryClientProvider>
    </MuiThemeProvider>
  );
};

export const renderComponent = (children: ReactElement, options?: Options) => {
  const queryClient = createQueryClient();

  const renderRes = renderComponentTl(children, {
    wrapper: props => <Wrapper queryClient={queryClient} options={options} {...props} />,
  });

  return {
    ...renderRes,
    queryClient,
  };
};

export const renderHook = <TProps, TResult>(
  hook: (props: TProps) => TResult,
  options?: Options,
) => {
  const queryClient = createQueryClient();

  const renderRes = renderHookTl(hook, {
    wrapper: props => <Wrapper queryClient={queryClient} options={options} {...props} />,
  });

  return {
    ...renderRes,
    queryClient,
  };
};
