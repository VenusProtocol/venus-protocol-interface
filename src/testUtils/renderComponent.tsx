import React from 'react';
import { render } from '@testing-library/react';
import { QueryClient, QueryClientProvider } from 'react-query';
import { BrowserRouter, Route, Switch } from 'react-router-dom';
import { ToastContainer } from 'react-toastify';

import { Web3Wrapper } from 'clients/web3';
import { AuthContext, IAuthContextValue } from 'context/AuthContext';
import { SuccessfulTransactionModalProvider } from 'context/SuccessfulTransactionModalContext';
import Theme from 'theme';
import { RefreshContextProvider } from 'context/RefreshContext';
import { VaiContext, IVaiContextValue } from 'context/VaiContext';
import { MuiThemeProvider } from 'theme/MuiThemeProvider/MuiThemeProvider';
import BigNumber from 'bignumber.js';

const renderComponent = (
  children: React.ReactElement | (() => React.ReactElement),
  {
    authContextValue = {},
    vaiContextValue = {},
  }: {
    authContextValue?: Partial<IAuthContextValue>;
    vaiContextValue?: Partial<IVaiContextValue>;
  } = {},
) => {
  const queryClient = new QueryClient({
    defaultOptions: {
      queries: {
        retry: false,
        cacheTime: 0,
      },
    },
  });
  const defaultAuthContextValues = {
    login: jest.fn(),
    logOut: jest.fn(),
    openAuthModal: jest.fn(),
    closeAuthModal: jest.fn(),
    account: undefined,
    ...authContextValue,
  };
  const defaultVaiContextValues = {
    userVaiMinted: new BigNumber(0),
    userVaiBalance: new BigNumber(0),
    userVaiEnabled: false,
    mintableVai: new BigNumber(0),
    ...vaiContextValue,
  };

  const renderRes = render(
    <Theme>
      <Web3Wrapper>
        <QueryClientProvider client={queryClient}>
          <MuiThemeProvider>
            <AuthContext.Provider value={defaultAuthContextValues}>
              <RefreshContextProvider>
                <VaiContext.Provider value={defaultVaiContextValues}>
                  <SuccessfulTransactionModalProvider>
                    <BrowserRouter>
                      <ToastContainer />

                      <Switch>
                        <Route
                          path="/"
                          component={typeof children === 'function' ? children : () => children}
                        />
                      </Switch>
                    </BrowserRouter>
                  </SuccessfulTransactionModalProvider>
                </VaiContext.Provider>
              </RefreshContextProvider>
            </AuthContext.Provider>
          </MuiThemeProvider>
        </QueryClientProvider>
      </Web3Wrapper>
    </Theme>,
  );

  return {
    ...renderRes,
    queryClient,
  };
};

export default renderComponent;
