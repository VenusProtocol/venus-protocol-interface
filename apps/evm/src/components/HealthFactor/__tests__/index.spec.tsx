import { renderComponent } from 'testUtils/render';
import { HealthFactor } from '..';
import { HEALTH_FACTOR_MAX_VALUE } from '../formatHealthFactorToReadableValue';

describe('HealthFactor', () => {
  it.each([
    Number.POSITIVE_INFINITY,
    HEALTH_FACTOR_MAX_VALUE + 1,
    12.23123,
    2.912312,
    1.989765,
    0.8738529,
  ])('displays correct UI for factor: %s', factor => {
    const { container: base } = renderComponent(<HealthFactor factor={factor} />);

    expect(base.textContent).toMatchSnapshot();
  });
});
