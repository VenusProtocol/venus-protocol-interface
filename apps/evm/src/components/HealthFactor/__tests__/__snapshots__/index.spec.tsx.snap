// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`HealthFactor > displays correct UI for factor: 0.8738529 1`] = `"0.87"`;

exports[`HealthFactor > displays correct UI for factor: 1.989765 1`] = `"1.98"`;

exports[`HealthFactor > displays correct UI for factor: 2.912312 1`] = `"2.91"`;

exports[`HealthFactor > displays correct UI for factor: 12.23123 1`] = `"12.23"`;

exports[`HealthFactor > displays correct UI for factor: 101 1`] = `"> 100"`;

exports[`HealthFactor > displays correct UI for factor: Infinity 1`] = `""`;
